Update EIP-5164: Add chainId and naming (#6146)
ethereum/EIPs: EIPS/eip-5164.md

This specification allows contracts on one chain to send messages to contracts on another chain. There are two key interfaces that needs to be implemented:

- `MessageDispatcher`
- `MessageExecutor`

The `MessageDispatcher` lives on the origin chain and dispatches messages to the `MessageExecutor` for execution. The `MessageExecutor` lives on the destination chain and executes dispatched messages.

There are also extensions of `MessageExecutor`, each defining a method for initiating a message or message batch:

- `SingleChainMessageDispatcher`
- `MultiChainMessageDispatcher`
- `BatchedMessageDispatcher`

Alternatively, `MessageDispatcher`s may implement a custom interface for initiating messages.

### MessageDispatcher

The `MessageDispatcher` lives on the chain from which messages are sent. The Dispatcher's job is to broadcast messages through a transport layer to one or more `MessageExecutor` contracts.

#### MessageDispatcher Events

> A unique `messageId` MUST be generated for each message or message batch.

Reviewer: I think we should lift this up into the MessageDispatcher description, and give some more detail.  It's going to be important for the messageId to be unique across contracts and chains- perhaps we can suggest that it's a monotonically increasing nonce

```solidity
interface MessageDispatcher {
  event MessageDispatched(
    bytes32 indexed messageId,
    address indexed from,
    uint256 indexed toChainId,
    address to,
    bytes data,
  );
}
```

```yaml
- name: MessageDispatched
  type: event
  inputs:
    - name: messageId
      indexed: true
      type: bytes32
    - name: from
      indexed: true
      type: address
    - name: toChainId
      indexed: true
      type: uint256
    - name: to
      type: address
    - name: data
      type: bytes
```

**MessageBatchDispatched**

The `MessageBatchDispatched` event MUST be emitted by the `MessageDispatcher` when a batch of messages is dispatched.

```solidity
struct Message {
  address to;
  bytes data;
}

interface MessageDispatcher {
  event MessageBatchDispatched(
    bytes32 indexed messageId,
    address indexed from,
    uint256 indexed toChainId,
    Message[] messages
  );
}
```

```yaml
- name: MessageBatchDispatched
  type: event
  inputs:
    - name: messageId
      indexed: true
      type: bytes32
    - name: from
      indexed: true
      type: address
    - name: toChainId
      indexed: true
      type: uint256
    - name: messages
      type: Message[]
```

> ### BasicMessageDispatcher

Reviewer: Can we rename this to SingleMessageDispatcher?

The `BasicMessageDispatcher` is an extension of `MessageDispatcher` that defines a method, `dispatchMessage`, for dispatching an individual message to be executed on the `toChainId`.

> #### Methods

Reviewer 2: ```suggestion
#### BasicMessageDispatcher Methods
```

**dispatchMessage**

Will dispatch a message to be executed by the `MessageExecutor` on the destination chain specified by `toChainId`.

`BasicMessageDispatcher`s MUST emit the `MessageDispatched` event when a message is dispatched.

> `BasicMessageDispatcher`s MUST forward the message to a `MessageExecutor` on the `toChainId`.

Reviewer: MUST revert if the chainId is unsupported

`BasicMessageDispatcher`s MUST use a unique `messageId` for each message.

> `BasicMessageDispatcher`s MUST return the `messageId` to allow the message sender to track the messages.

Reviewer 2: ```suggestion
`BasicMessageDispatcher`s MUST return the `messageId` to allow the message sender to track the message.
```

`BasicMessageDispatcher`s MAY require payment.

```solidity
interface BasicMessageDispatcher is MessageDispatcher {
  function dispatchMessage(uint256 toChainId, address to, bytes calldata data) external payable returns (bytes32 messageId);
}
```

```yaml
- name: dispatchMessage
  type: function
  stateMutability: payable
  inputs:
    - name: toChainId
      type: uint256
    - name: to
      type: address
    - name: data
      type: bytes
  outputs:
    - name: messageId
      type: bytes32
```

### BatchedMessageDispatcher

The `BatchedMessageDispatcher` is an extension of `MessageDispatcher` that defines a method, `dispatchMessageBatch`, for dispatching a batch of messages to be executed on the `toChainId`.

> #### Methods

Reviewer 2: ```suggestion
#### BatchedMessageDispatcher Methods
```

**dispatchMessageBatch**

Will dispatch a batch of messages to be executed by the `MessageExecutor` on the destination chain specified by `toChainId`.

> `MultiChainMessageDispatcher`s MUST emit the `MessageBatchDispatched` event when a message batch is dispatched.

Reviewer: MUST revert if it's an unsupported chainId

> `MultiChainMessageDispatcher`s MUST forward the message batch to the `MessageExecutor` on the `toChainId`.

Reviewer: Rename MultiChainMessageDispatcher to BatchedMessageDispatcher

`MultiChainMessageDispatcher`s MUST use a unique `messageId` for each batch of messages.

> `MultiChainMessageDispatcher`s MUST return the `messageId` to allow the message sender to track the message batches.

Reviewer 2: ```suggestion
`MultiChainMessageDispatcher`s MUST return the `messageId` to allow the message sender to track the batch of messages.
```

> `MultiChainMessageDispatcher`s MUST emit the `MessageBatchDispatched` event when a message batch is dispatched.
> 
> `MultiChainMessageDispatcher`s MUST forward the message batch to the `MessageExecutor` on the `toChainId`.
> 
> `MultiChainMessageDispatcher`s MUST use a unique `messageId` for each batch of messages.
> 
> `MultiChainMessageDispatcher`s MUST return the `messageId` to allow the message sender to track the message batches.
> 
> `MultiChainMessageDispatcher`s MAY require payment.

Reviewer 2: Should be `BatchedMessageDispatcher`:
```suggestion
`BatchedMessageDispatcher`s MUST emit the `MessageBatchDispatched` event when a message batch is dispatched.

`BatchedMessageDispatcher`s MUST forward the message batch to the `MessageExecutor` on the `toChainId`.

`BatchedMessageDispatcher`s MUST use a unique `messageId` for each batch of messages.

`BatchedMessageDispatcher`s MUST return the `messageId` to allow the message sender to track the message batches.

`BatchedMessageDispatcher`s MAY require payment.
```

```solidity

> interface MultiChainMessageDispatcher is MessageDispatcher {

Reviewer 2: ```suggestion
interface BatchedMessageDispatcher is MessageDispatcher {
```

  function dispatchMessageBatch(uint256 toChainId, Message[] calldata messages) external payable returns (bytes32 messageId);
}
```

```yaml

> - name: dispatchMessage

Reviewer 2: ```suggestion
- name: dispatchMessageBatch
```

  type: function
  stateMutability: payable
  inputs:
    - name: toChainId
      type: uint256
    - name: messages
      type: Message[]
  outputs:
    - name: messageId
      type: bytes32
```

### MessageExecutor

The `MessageExecutor` executes dispatched messages and message batches. Developers must implement a `MessageExecutor` in order to execute messages on the receiving chain.

The `MessageExecutor` will execute a messageId only once, but may execute messageIds in any order. This specification makes no ordering guarantees, because messages and message batches may travel non-sequentially through the transport layer.

#### Execution

`MessageExecutor`s SHOULD verify all message data with the bridge transport layer.

`MessageExecutor`s MUST NOT successfully execute a message more than once.

`MessageExecutor`s MUST revert the transaction when a message fails to be executed allowing the message to be retried at a later time.

**Calldata**

> `MessageExecutor`s MUST append the ABI-packed (`fromChainId`, `from`) to the calldata for each message being executed. This allows the receiver of the message to verify the cross-chain sender and the chain that the message is coming from.

Reviewer 2: Should we also pass the `messageId` to allow the receiving contract to handle/verify transaction ordering?
That's why we were packing the `nonce` in the first place.

Author: Yeah, sounds good. It will be useful for pairing the execution event with external events.

```solidity
interface MessageExecutor {

>   bytes calldata = abi.encodePacked(fromChainId, from);

Reviewer 2: We also need to pass the data in here:
```suggestion
  bytes calldata = abi.encodePacked(data, fromChainId, from);
```

Reviewer: This code snippet feels strange because it's not valid solidity... maybe exclude the wrapping `interface` declaration? I feel like my brain wouldn't try to compile it then.

something like:

```
to.call(abi.encodePacked(_call.data, _nonce, _sender));
```

Author: @PierrickGT good catch, didn't mean to remove that

@asselstine sounds good to me. Updated

}
```

```yaml
- name: calldata
  type: bytes
  inputs:

>     - name: fromChainId

Reviewer 2: ```suggestion
    - name: data
       type: bytes
    - name: fromChainId
```

      type: uint256
    - name: from
      type: address
```

#### Error handling

**MessageAlreadyExecuted**

`MessageExecutor`s MUST revert if a messageId has already been executed and SHOULD emit a `MessageIdAlreadyExecuted` custom error.

```solidity
interface MessageExecutor {
  error MessageIdAlreadyExecuted(
    bytes32 messageId
  );
}
```

**MessageFailure**

`MessageExecutor`s MUST revert if an individual message fails and SHOULD emit a `MessageFailure` custom error.

```solidity
interface MessageExecutor {
  error MessageFailure(
    bytes errorData
  );
}
```

**MessageBatchFailure**

`MessageExecutor`s MUST revert the entire batch if any message in a batch fails and SHOULD emit a `MessageBatchFailure` custom error.

```solidity
interface MessageExecutor {
  error MessageBatchFailure(
    uint256 messageIndex,
    bytes errorData
  );
}
```

#### MessageExecutor Events

**MessageIdExecuted**

`MessageIdExecuted` MUST be emitted once a message or message batch has been executed.

```solidity
interface MessageExecutor {

>   event MessageIdExecuted(

Reviewer: Suggest renaming to "MessageExecuted" and excluding the `fromChainId`

Indexers will see the "MessageDispatched" first on the sending chain, so they'll already have all the info

Author: The `fromChainId` is useful in frontend/sdk use cases where you have the `MessageExecuted` event and want to look up the `MessageDispatched` event back on the origin chain. That way you don't need to check against every network you support when things like subgraphs, RPCs, etc. are divided by network.

    uint256 indexed fromChainId,
    bytes32 indexed messageId
  );
}
```

```yaml

> - name: MessageIdExecuted

Reviewer: Same same

`MessageExecutor`s MUST revert the transaction when a message fails to be executed allowing the message to be retried at a later time.

**Calldata**

`MessageExecutor`s MUST append the ABI-packed (`messageId`, `fromChainId`, `from`) to the calldata for each message being executed. This allows the receiver of the message to verify the cross-chain sender and the chain that the message is coming from.

```solidity
to.call(abi.encodePacked(data, messageId, fromChainId, from));
```

```yaml
- name: calldata
  type: bytes
  inputs:
    - name: data
      type: bytes
    - name: messageId
      type: bytes32
    - name: fromChainId
      type: uint256
    - name: from
      type: address
```

#### Error handling

**MessageAlreadyExecuted**

`MessageExecutor`s MUST revert if a messageId has already been executed and SHOULD emit a `MessageIdAlreadyExecuted` custom error.

```solidity
interface MessageExecutor {
  error MessageIdAlreadyExecuted(
    bytes32 messageId
  );
}
```

**MessageFailure**

`MessageExecutor`s MUST revert if an individual message fails and SHOULD emit a `MessageFailure` custom error.

```solidity

> interface MessageExecutor {
>   error MessageFailure(
>     bytes errorData
>   );
> }

Reviewer 2: ```suggestion
interface MessageExecutor {
  error MessageFailure(
    bytes32 messageId,
    bytes errorData
  );
}
```

```

**MessageBatchFailure**

`MessageExecutor`s MUST revert the entire batch if any message in a batch fails and SHOULD emit a `MessageBatchFailure` custom error.

```solidity

> interface MessageExecutor {
>   error MessageBatchFailure(
>     uint256 messageIndex,
>     bytes errorData
>   );
> }

Reviewer 2: ```suggestion
interface MessageExecutor {
  error MessageBatchFailure(
    bytes32 messageId,
    uint256 messageIndex,
    bytes errorData
  );
}
```

```

#### MessageExecutor Events

**MessageIdExecuted**

`MessageIdExecuted` MUST be emitted once a message or message batch has been executed.

```solidity
interface MessageExecutor {
  event MessageIdExecuted(
    uint256 indexed fromChainId,
    bytes32 indexed messageId
  );
}
```

```yaml
- name: MessageIdExecuted
  type: event
  inputs:
    - name: fromChainId
      indexed: true
      type: uint256
    - name: messageId
      indexed: true
      type: bytes32
```

## Rationale

> The `MessageDispatcher` can be coupled to one or more `MessageExecutor`. It is up to bridges to decide how to couple the two. Users can easily bridge a message by calling `dispatchMessage` without being aware of the `MessageExecutor` address. Messages can also be traced by a client using the data logged by the `ExecutedMessage` event.

Reviewer 2: ```suggestion
The `MessageDispatcher` can be coupled to one or more `MessageExecutor`. It is up to bridges to decide how to couple the two. Users can easily bridge a message by calling `dispatchMessage` without being aware of the `MessageExecutor` address. Messages can also be traced by a client using the data logged by the `MessageIdExecuted` event.
```

#### Methods

**relayCall**

Will relay a single `Call` struct as a batch of one to be executed by any `CrossChainExecutor`(s) that executes the calls from this Relayer.

`CrossChainRelayer`s MUST emit the `RelayedCalls` event when a batch of calls is relayed.

> `CrossChainRelayer`s MAY ignore the `toChainId` when there is one or more predetermined destination chain(s).

Reviewer 2: Users may want to relay a batch of calls to one or more chains, so I'm wondering if `toChainId` should be a `uint256` array?

> `CrossChainRelayer`s SHOULD forwarding a batch to the `CrossChainExecutor`s on the `toChainId`.

Reviewer 2: ```suggestion
`CrossChainRelayer`s SHOULD forward a batch to the `CrossChainExecutor`s on the `toChainId`.
```

`CrossChainRelayer`s MUST increment a `nonce` so that each batch of calls can be uniquely identified.

`CrossChainRelayer`s MUST return the `nonce` to allow callers to track the batch of calls.

`CrossChainRelayer`s SHOULD pass the `nonce` as well as the `from` address in the call to `CrossChainExecutor` to uniquely identify the batch of calls on the receiving chain.

`CrossChainRelayer`s MAY require payment.

```solidity
interface CrossChainRelayer {
  function relayCall(Call calldata call, uint256 toChainId) external payable returns (uint256 nonce);
}
```

```yaml
- name: relayCall
  type: function
  stateMutability: payable
  inputs:
    - name: call
      type: Call
    - name: toChainId
      type: uint256
  outputs:
    - name: nonce
      type: uint256
```

**relayCalls**

Will relay a batch of `Call` structs to be executed by any `CrossChainExecutor`(s) that execute calls from this Relayer.

`CrossChainRelayer`s MUST emit the `RelayedCalls` event when a batch of calls is relayed.

`CrossChainRelayer`s MAY ignore the `toChainId` when there is one or more predetermined destination chain(s).

> `CrossChainRelayer`s SHOULD forwarding a batch to the `CrossChainExecutor`s on the `toChainId`.

Reviewer 2: ```suggestion
`CrossChainRelayer`s SHOULD forward a batch to the `CrossChainExecutor`s on the `toChainId`.
```
